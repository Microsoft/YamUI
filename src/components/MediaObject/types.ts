/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export enum MediaObjectSize {
  XLARGE = 'xLarge',
  LARGE = 'large',
  MEDIUM = 'medium',
  SMALL = 'small',
  XSMALL = 'xSmall',
}
