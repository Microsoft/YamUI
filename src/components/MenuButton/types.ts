/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { ContextualMenuItemType } from 'office-ui-fabric-react/lib/ContextualMenu';

export enum MenuItemType {
  Normal = ContextualMenuItemType.Normal,
  Header = ContextualMenuItemType.Header,
  Divider = ContextualMenuItemType.Divider,
}
