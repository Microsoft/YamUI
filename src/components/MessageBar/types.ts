/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export enum MessageBarType {
  INFO = 'info',
  WARNING = 'warning',
  ERROR = 'error',
}
