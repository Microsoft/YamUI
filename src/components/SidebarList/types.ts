/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { BaseIllustration } from '../Illustration';

export type SidebarListSize = 'small' | 'medium' | 'large';

export interface BaseSidebarListItem {
  /**
   * A unique key for the item.
   */
  key: string;

  /**
   * The visible title text.
   */
  title: string;

  /**
   * Metadata text displayed under the title, not displayed at small size.
   */
  description?: string;
}
export interface SidebarListIllustrationItem extends BaseSidebarListItem {
  /**
   * A YamUI Illustration component class. You are responsible for importing and providing the correct size.
   */
  illustration: typeof BaseIllustration;
}
export interface SidebarListImageItem extends BaseSidebarListItem {
  /**
   * Custom image JSX. It should fit the list item's pixel width at the given size, or fill 100% of its container width.
   */
  imageContent: React.ReactNode;
}
export type SidebarListItemWithGraphic = SidebarListIllustrationItem | SidebarListImageItem;

export type SidebarListClickableItem = SidebarListItemWithGraphic & {
  /**
   * Callback to be triggered when the item is clicked/invoked.
   */
  onClick(key: BaseSidebarListItem['key']): void;
};
export type SidebarListLinkItem = SidebarListItemWithGraphic & {
  /**
   * The URL this item links to.
   */
  href: string;

  /**
   * If true, opens the link securely in a new window.
   */
  newWindow?: boolean;
};
export type SidebarListItem = SidebarListClickableItem | SidebarListLinkItem;
