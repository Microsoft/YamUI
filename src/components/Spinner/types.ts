/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
export enum SpinnerColor {
  /**
   * Light theme.
   */
  LIGHT = 'light',
  /**
   * Light theme spinner color with gray text.
   */
  METADATA = 'metadata',
  /**
   * Dark theme.
   */
  DARK = 'dark',
}
export enum SpinnerSize {
  LARGE = 'large',
  MEDIUM = 'medium',
  SMALL = 'small',
  XSMALL = 'xSmall',
}
