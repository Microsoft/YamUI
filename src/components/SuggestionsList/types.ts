/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */

import { AvatarProps } from '../Avatar';

export enum SuggestionsListItemTemplate {
  TEXT = 'text',
  AVATAR_ROUND = 'mediaRound',
  MEDIA_SOFT = 'mediaSoft',
  MEDIA_SQUARE = 'mediaSquare',
}

export type CustomizableAvatarProps = 'badgeContent' | 'badgeDescription';

export interface BaseSuggestionItem {
  id: string | number;
  name: string;
}
export interface TextSuggestionItem extends BaseSuggestionItem {
  template: SuggestionsListItemTemplate.TEXT;
}
export interface MediaSuggestionItem extends BaseSuggestionItem {
  template: SuggestionsListItemTemplate.MEDIA_SOFT | SuggestionsListItemTemplate.MEDIA_SQUARE;
  imageUrl?: string;
  description?: string;
}
export interface AvatarSuggestionItem extends BaseSuggestionItem {
  template: SuggestionsListItemTemplate.AVATAR_ROUND;
  imageUrl?: string;
  avatarProps?: Pick<AvatarProps, CustomizableAvatarProps>;
  description?: string;
}
export type SuggestionItem = TextSuggestionItem | MediaSuggestionItem | AvatarSuggestionItem;
