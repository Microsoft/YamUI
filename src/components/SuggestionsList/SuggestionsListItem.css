/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
@import "src/css/variables/colors.css";

.y-suggestions-list-item {
  cursor: pointer;
  padding: 0.4rem 1.2rem 0.8rem 1.2rem;
}

.y-suggestions-list-item:hover {
  background-color: $suggestionsListItemBackgroundColor__mouseHover;
}

.y-suggestions-list-item--selected {
  background-color: $suggestionsListItemBackgroundColor__keyboardNavigate;
}

.y-suggestions-list-item--highlight {
  text-decoration: underline;
}
