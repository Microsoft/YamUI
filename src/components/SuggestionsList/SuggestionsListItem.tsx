/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';
import MediaObject, { MediaObjectSize } from '../MediaObject';
import Avatar, { AvatarProps, AvatarSize } from '../Avatar';

import './SuggestionsListItem.css';

export type CustomizableAvatarProps = 'badgeContent' | 'badgeDescription' | 'borderType';

export interface SuggestionItem {
  id: string | number;
  imageUrl?: string;
  name: string;
  description?: string;
  avatarProps?: Pick<AvatarProps, CustomizableAvatarProps>;
}

export interface SuggestionsListItemProps extends SuggestionItem, NestableBaseComponentProps {
  searchText: string;
  isSelected: boolean;
  onSelect(id: string | number): void;
}

const baseClass = 'y-suggestions-list-item y-hc-select-on-hover y-hc-suppress-text-background';
const selectedClass = `${baseClass} y-suggestions-list-item--selected y-hc-selected`;
const matchHighlightClass = 'y-suggestions-list-item--match-highlight';

const getHighlightedName = (name: string, search: string) => {
  return name
    .split(new RegExp(`(${search})`, 'gi'))
    .filter(content => !!content)
    .map((content, index) => {
      const className =
        search.toLowerCase() === content.toLowerCase() ? matchHighlightClass : undefined;
      return (
        <span key={index} className={className}>
          {content}
        </span>
      );
    });
};

export default class SuggestionsListItem extends React.PureComponent<SuggestionsListItemProps, {}> {
  public render() {
    const { isSelected, name, searchText, description } = this.props;
    const avatar = this.getAvatar();
    const className = isSelected ? selectedClass : baseClass;
    const title = getHighlightedName(name, searchText);
    return (
      <div onMouseDown={this.onMouseDown} className={className}>
        <MediaObject
          size={MediaObjectSize.SMALL}
          imageContent={avatar}
          titleContent={title}
          metadataContent={description}
        />
      </div>
    );
  }

  private getAvatar() {
    const { avatarProps, name, imageUrl } = this.props;
    const props = {
      ...avatarProps,
      name,
      imageUrl,
      size: AvatarSize.SMALL,
    };
    return <Avatar {...props} />;
  }

  private onMouseDown = () => {
    const { onSelect, children, className, ...item } = this.props;
    this.props.onSelect(item.id);
  };
}
