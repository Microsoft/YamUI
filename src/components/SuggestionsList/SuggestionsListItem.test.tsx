/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import SuggestionsListItem, { SuggestionsListItemProps } from './SuggestionsListItem';
import { SuggestionItem, SuggestionsListItemTemplate } from './types';

describe('<SuggestionsListItem />', () => {
  let component: ShallowWrapper<SuggestionsListItemProps>;
  let onSelect: jest.Mock;

  const textItem: SuggestionItem = {
    template: SuggestionsListItemTemplate.TEXT,
    id: '1',
    name: 'John Smith',
  };

  const getProps = (overrides?: Partial<SuggestionsListItemProps>) => {
    onSelect = jest.fn();

    return {
      onSelect,
      isSelected: false,
      searchText: 'Joh',
      item: textItem,
      ...overrides,
    };
  };

  beforeEach(() => {
    onSelect = jest.fn();
  });

  describe('when isSelected=true', () => {
    beforeEach(() => {
      component = shallow(<SuggestionsListItem {...getProps({ isSelected: true })} />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('when selecting', () => {
    let preventDefault: jest.Mock;

    beforeEach(() => {
      component = shallow(<SuggestionsListItem {...getProps()} />);
      preventDefault = jest.fn();
      component.simulate('mouseDown', { preventDefault });
    });

    it('calls props.onSelect', () => {
      expect(onSelect).toBeCalled();
    });

    it('calls event preventDefault', () => {
      expect(preventDefault).toBeCalled();
    });
  });
});
