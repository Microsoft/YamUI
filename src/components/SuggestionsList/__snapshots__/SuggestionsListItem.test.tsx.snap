// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SuggestionsListItem /> when isSelected=false renders as expected 1`] = `
<div
  className=
      y-suggestions-list-item
      y-hc-select-on-hover
      y-hc-suppress-text-background
      {
        cursor: pointer;
        padding-bottom: 0.8rem;
        padding-left: 1.2rem;
        padding-right: 1.2rem;
        padding-top: 0.4rem;
      }
      &:hover {
        background-color: #f3f5f8;
      }
  onMouseDown={[Function]}
  role="button"
>
  <MediaObject
    imageContent={
      <Avatar
        imageUrl="imageUrl"
        name="John Smith"
        size="small"
      />
    }
    metadataContent="Software Engineer"
    size="small"
    titleContent={
      Array [
        <span
          className=

              {
                font-weight: 600;
              }
        >
          Joh
        </span>,
        <span>
          n Smith
        </span>,
      ]
    }
  />
</div>
`;

exports[`<SuggestionsListItem /> when isSelected=true renders as expected 1`] = `
<div
  className=
      y-suggestions-list-item
      y-hc-select-on-hover
      y-hc-suppress-text-background
      y-hc-selected
      {
        background-color: #edeff2;
        cursor: pointer;
        padding-bottom: 0.8rem;
        padding-left: 1.2rem;
        padding-right: 1.2rem;
        padding-top: 0.4rem;
      }
      &:hover {
        background-color: #f3f5f8;
      }
  onMouseDown={[Function]}
  role="button"
>
  <MediaObject
    imageContent={
      <Avatar
        imageUrl="imageUrl"
        name="John Smith"
        size="small"
      />
    }
    metadataContent="Software Engineer"
    size="small"
    titleContent={
      Array [
        <span
          className=

              {
                font-weight: 600;
              }
        >
          Joh
        </span>,
        <span>
          n Smith
        </span>,
      ]
    }
  />
</div>
`;

exports[`<SuggestionsListItem /> with avatarProps renders as expected 1`] = `
<div
  className=
      y-suggestions-list-item
      y-hc-select-on-hover
      y-hc-suppress-text-background
      {
        cursor: pointer;
        padding-bottom: 0.8rem;
        padding-left: 1.2rem;
        padding-right: 1.2rem;
        padding-top: 0.4rem;
      }
      &:hover {
        background-color: #f3f5f8;
      }
  onMouseDown={[Function]}
  role="button"
>
  <MediaObject
    imageContent={
      <Avatar
        borderType="soft"
        imageUrl="imageUrl"
        name="John Smith"
        size="small"
      />
    }
    metadataContent="Software Engineer"
    size="small"
    titleContent={
      Array [
        <span
          className=

              {
                font-weight: 600;
              }
        >
          Joh
        </span>,
        <span>
          n Smith
        </span>,
      ]
    }
  />
</div>
`;
